src/testers.rs

```diff
@@ -1,30 +1,130 @@
+fn number_args_allowed(arg_name: &str, tester_name: &str, max: usize, args_len: usize) -> TeraResult<()> {
+    if max == 0 && args_len > max {
+        return Err(TeraError::TestError(
+            tester_name.to_string(),
+            format!(
+                "{} was called on the variable {} with some arguments \
+                but this test doesn't take arguments.",
+                tester_name, arg_name
+            )
+        ))
+    }
+
+    if args_len > max {
+        return Err(TeraError::TestError(
+            tester_name.to_string(),
+            format!(
+                "{} was called on the variable {} with {} arguments, the max number is {}. ",
+                tester_name, arg_name, args_len, max
+            )
+        ))
+    }
+
+    Ok(())
+}
+
+// Called to return an error when unwrapping the value and realising there's nothing
+fn value_is_undefined(arg_name: &str, tester_name: &str) -> TeraResult<bool> {
+    Err(TeraError::TestError(
+        tester_name.to_string(),
+        format!("{} was called on the variable {}, which is undefined", tester_name, arg_name)
+    ))
+}
+
+/// Returns true if `value` is defined. Otherwise, returns false.
+pub fn defined(name: &str, value: Option<Value>, params: Vec<Value>) -> TeraResult<bool> {
+    try!(number_args_allowed(name, "defined", 0, params.len()));
+
+    Ok(value.is_some())
+}
+
+/// Returns true if `value` is undefined. Otherwise, returns false.
+pub fn undefined(name: &str, value: Option<Value>, params: Vec<Value>) -> TeraResult<bool> {
+    try!(number_args_allowed(name, "undefined", 0, params.len()));
+
+    Ok(value.is_none())
+}
+
+/// Returns true if `value` is a string. Otherwise, returns false.
+pub fn string(name: &str, value: Option<Value>, params: Vec<Value>) -> TeraResult<bool> {
+    try!(number_args_allowed(name, "string", 0, params.len()));
+
+    match value {
+        Some(Value::String(_)) => Ok(true),
+        None => value_is_undefined(name, "string"),
+        _ => Ok(false)
     }
+}
 
-    let name = match *value {
-        Node::Identifier { ref name, .. } => name,
-        _ => return Err(TeraError::TestError("defined".to_string(),
-                "defined can only be called on identifiers".to_string()))
-    };
+/// Returns true if `value` is a number. Otherwise, returns false.
+pub fn number(name: &str, value: Option<Value>, params: Vec<Value>) -> TeraResult<bool> {
+    try!(number_args_allowed(name, "number", 0, params.len()));
 
-    Ok(context.contains_key(name))
+    match value {
+        Some(Value::I64(_)) | Some(Value::F64(_)) | Some(Value::U64(_)) => Ok(true),
+        None => value_is_undefined(name, "number"),
+        _ => Ok(false)
+    }
 }
 
+/// Returns true if `value` is an odd number. Otherwise, returns false.
+pub fn odd(name: &str, value: Option<Value>, params: Vec<Value>) -> TeraResult<bool> {
+    try!(number_args_allowed(name, "odd", 0, params.len()));
+
+    match value {
+        Some(v) => match v.to_number() {
```

Comment:
I think the `value_is_undefined` call should be renamed `value_defined` and used as in a try expression. Then, this function (and others) can be rewritten like:

```rust
try!(number_args_allowed(name, "odd", 0, params.len()));
try!(value_defined(name, "odd"));

match value.and_then(|v| v.to_number()) {
    Some(number) => Ok(number % 2.0 != 0.0),
    None => return Err(TeraError::TestError(
        "odd".into(), "odd can only be called on numbers".into())
    ),
}
```

I think this looks significantly cleaner.